### PR TITLE
Update outdated pins & ignore star pinnings

### PR DESCRIPTION
**Pull request by jschueller**
- Update outdated pins: enhances #310 with code from https://github.com/croth1/conda-forge-pinning/pull/3
- Ignore star pinnings: implements #270
cc @jakirkham @krischer @croth1

### CODE REVIEW

**Comment by ocefpaf** [comment]
@jschueller while I appreciate the effort to fix our pinning we should really invest time in moving to conda-build 3. This will be obsolete soon.

**Comment by jschueller** [comment]
: /
Can you set me up to speed with the conda-build 3 stuff ?

**Comment by jschueller** [comment]
I mean what's changed with conda build 3? how does it affect pinnings? what's left to do to use conda build 3?

**Comment by jakirkham** [comment]
Really have lost track. The goal some time back was to have a package with our pinnings that would work on `conda-build` 2 and 3. There were some fixes needed to it. Please see PR ( https://github.com/conda-forge/staged-recipes/pull/2267 ) for details.

**Comment by ocefpaf** [comment]
Unfortunately things are slow mostly due to our slow pace to implements things.
But here is where we are at the moment:

- we don't plan to start using conda-build 3 in `conda-smithy` at the moment, we'll do like the cb1 to cb2 change by hardcoding it the the CIs until we get some key packages correctly.
- we'll have a centralized config file with the pinning info (check conda-build 3 docs for more info)
- there will be virtually no pinning in the recipes; things will be defined in this common config file

BTW, take a look at https://github.com/conda-forge/geos-feedstock/pull/25 and let's move this discussion there.

**Comment by isuruf** [comment]
> we'll have a centralized config file with the pinning info (check conda-build 3 docs for more info)

I don't understand this. Say we pin a package to one version in the config file, and when we update the version, how would we know which conda packages we need to rebuild?

**Comment by ocefpaf** [comment]
> I don't understand this. Say we pin a package to one version in the config file, and when we update the version, how would we know which conda packages we need to rebuild?

We don't. When updating the pinning we'll have to relying on re-rendering to get the new packages out.

We can improve the re-renderings to have some understanding about the dependency graph and re-build them in order. (Which is a big improvement to the current situation even if the current situation wasn't broken ;-)

**Comment by jschueller** [comment]
Maybe the pinning & rerendering bot can still be useful until we get there ?

**Comment by ocefpaf** [comment]
> Maybe the pinning & rerendering bot can still be useful until we get there ?

I am not against it! Please if you want to fix this go ahead and it will be good for us.

I just want to warn you that we should be focusing on moving to conda-build 3 ASAP.

**Comment by isuruf** [comment]
> We don't. When updating the pinning we'll have to relying on re-rendering to get the new packages out.

Do you mean we need to rerender all 3000+ packages when we update one pinning?

**Comment by ocefpaf** [comment]
> Do you mean we need to rerender all 3000+ packages when we update one pinning?

If you know the dependency graph no.

BTW. It is just a matter of adjusting the bot send the re-render with a build number bump to package that depend on something that is in the pinning config. Not much different from now, it will never be 3000+

**Comment by isuruf** [comment]
Creating the dependency graph of all packages. That's going to be fun.

**Comment by ocefpaf** [comment]
> Creating the dependency graph of all packages. That's going to be fun.

Actually that is fun :smile: 

@ericdill has a nice way to do that using the json file from a conda-env.

I have a few ideas for some new functionalities in conda-smithy and the bots using that for the next Google Summer of Code. Let's hope we can get a student (shameless pinging @kain88-de :stuck_out_tongue_winking_eye:)

**Comment by isuruf** [comment]
I guess then it's going to be a looong time before we can upgrade to conda-build 3. Look at how much time it took for us to use `noarch: python`. So, I guess fixing this in the meantime will be good.

**Comment by ocefpaf** [comment]
> So, I guess fixing this in the meantime will be good.

Like I said above, yes.

> I guess then it's going to be a looong time before we can upgrade to conda-build 3.

I really hope not. We cannot wait until shit hits the fan like the conda-build 1 to 2 move.

> Look at how much time it took for us to use `noarch: python`.

That was a mistake IMO due to lack of communication. We must strive to improve that.

**Comment by kain88-de** [comment]
I wanted to have a dependency graph project already this year for GSoC. I'll let you know once google announces GSoC for 2018.

**Comment by jakirkham** [comment]
Would reach out to bioconda about dependency graph code. They may have some ideas.

cc @johanneskoester @bgruening

**Comment by ocefpaf** [comment]
>  wanted to have a dependency graph project already this year for GSoC. 

That was my bad and not the applicants. I wrote it in the ideas doc locally but uploaded an older version :sob:

**Comment by ericdill** [comment]
Possibly related: [neat visualization of the nodejs ecosystem](https://anvaka.github.io/allnpmviz3d/)

**Comment by ericdill** [comment]
> @ericdill has a nice way to do that using the json file from a conda-env.

https://gist.github.com/ericdill/9942ac55c2c9f6a550973dd2dc3653a4

**Comment by bgruening** [comment]
@kain88-de @jakirkham bioconda has some code for this already in bioconda-utils. Please have a look at: https://github.com/bioconda/bioconda-utils/blob/master/bioconda_utils/utils.py#L310 and related code.

**Comment by jakirkham** [comment]
@krischer, could you please take a look? Want to make sure this change isn't going to break things for you.

**Comment by jschueller** [comment]
ping @krischer

File: scripts/pin_the_slow_way.py [hunk 1]
```diff
@@ -306,7 +306,12 @@ def create_update_pr(clone, remote_head, fork_remote, upstream_remote):
 
                 for pos, dep in enumerate(section):
                     for name, pin in pinned.items():
-                        if re.match(r"^\s*%s\s*$" % name, dep) and dep != pin:
+                        dep_split = dep.split(' ', 1)
+                        actual_name = dep_split[0]
+                        # skip star pins: https://github.com/conda-forge/conda-forge.github.io/issues/270
+                        if len(dep_split) > 1 and dep_split[1] == '*':
+                            continue
```
Comment (krischer) [inline scripts/pin_the_slow_way.py]: It would IMHO be clearer to also handle this with a regex but I also don't immediately see any flaws in this logic.

Comment (croth1) [reply scripts/pin_the_slow_way.py]: I had unittests in one of the stale PRs: https://github.com/conda-forge/conda-smithy/pull/482/files#diff-c81609e53a452e32dfd828bd02e72967.

Comment (krischer) [reply scripts/pin_the_slow_way.py]: Ah nice - might be worthwhile to cherry pick those on top of this branch here or to a separate branch.

**Comment by jakirkham** [comment]
Testing things is a great idea @krischer. Though it is a bit challenging with these scripts. Moving things into `conda-smithy` or the webservices generally makes things easier to test. @croth1 did a great job at working towards the former with pinnings. Unfortunately things are a bit in flux with `conda-build` 3. So we will need to revisit how we do pinnings from there we can revisit tests if they should be appropriate.

**Comment by jakirkham** [comment]
Thanks everyone for the great discussion and thanks @jschueller for proposing this. Going to go ahead and get this one in.

**Pull request merged**